src/macros.jl

```diff
@@ -1005,34 +1005,128 @@ function parsevariable(_error::Function, infoexpr::VariableInfoExpr, lvalue, lsi
+
+* `Bin`: Sets the variable to be binary, i.e. either 0 or 1.
+* `Int`: Sets the variable to be integer, i.e. one of ..., -2, -1, 0, 1, 2, ...
+* `Symmetric`: Only available when creating a square matrix of variables, i.e.
+  when `varexpr` is of the form `varname[1:n,1:n]` or `varname[i=1:n,j=1:n]`.
+  It creates a symmetric matrix of variable, that is, it only creates a
+  new variable for `varname[i,j]` with `i ≤ j` and sets `varname[j,i]` to the
+  same variable as `varname[i,j]`.
+* `PSD`: Same as `Symmetric` but also constrains the matrix to be positive
+  semidefinite.
+
+The recognized keyword arguments in `kwargs` are the following:
+
+* `basename`: Sets the base name used to generate variable names. It
+  corresponds to the variable name for scalar variable, otherwise, the
+  variable names are `basename[...]` for each indices `...` of the axes `axes`.
+* `lowerbound`: Sets the value of the variable lower bound.
+* `upperbound`: Sets the value of the variable upper bound.
+* `start`: Sets the variable starting value used as initial guess in optimization.
+* `binary`: Sets whether the variable is binary or not.
+* `integer`: Sets whether the variable is integer or not.
+* `variabletype`: See the "Note for extending the variable macro" section below.
+* `container`: Specify the container type, see [Containers in macro](@ref).
+
+## Note for extending the variable macro
+
+The single scalar variable or each scalar variable of the container are created
+using `addvariable(m, buildvariable(_error, info, extra_args...;
+extra_kwargs...))` where
+
+* `m` is the model passed to the `@variable` macro;
+* `_error` is an error function with a single `String` argument showing the
+  `@variable` call in addition to the error message given as argument;
+* `info` is the `VariableInfo` struct containing the information gathered in
+  `expr`, the recognized keyword arguments (except `basename` and
+  `variabletype`) and the recognized positional arguments (except `Symmetric`
+  and `PSD`);
+* `extra_args` are the unrecognized positional arguments of `args` plus the
+  value of the `variabletype` keyword argument if present. The `variabletype`
+  keyword argument allows the user to pass a position argument to
+  `buildvariable` without the need to give a positional argument to
+  `@variable`. In particular, this allows the user to give a positional
+  argument to the `buildvariable` call when using the anonymous single variable
+  syntax `@variable(m; kwargs...)`; and
+* `extra_kwargs` are the unrecognized keyword argument of `kwargs`.
+
+## Examples
+
+The following are equivalent ways of creating a variable `x` of name `x` with
+lowerbound 0:
+```julia
+# Specify everything in `expr`
+@variable(m, x >= 0)
+# Specify the lower bound using a keyword argument
+@variable(m, x, lowerbound=0)
+# Specify everything in `kwargs`
+x = @variable(m, basename="x", lowerbound=0)
+# Without the `@variable` macro
```

Comment:
Maybe put the simple examples that are relevant to most users first (before "Notes for extending ...") and move the examples for "Without the `@variable` macro" under the "Notes for extending ...". This is very advanced and may confuse users just looking for examples of how to use the macro.